libsolidity/ast/TypeProvider.h

```diff
@@ -0,0 +1,228 @@
+	TypeProvider& operator=(TypeProvider&&) = default;
+	TypeProvider& operator=(TypeProvider const&) = delete;
+	~TypeProvider() = default;
+
+	/**
+	 * Resets state of this TypeProvider to initial state, wiping all mutable types.
+	 *
+	 * This invalidates all dangling pointers to types provided by this TypeProvider.
+	 */
+	static void reset();
+
+	/// @name Factory functions
+	/// Factory functions that convert an AST @ref TypeName to a Type.
+	static Type const* fromElementaryTypeName(ElementaryTypeNameToken const& _type);
+
+	/// Converts a given elementary type name with optional data location
+	/// suffix " storage", " calldata" or " memory" to a type pointer. If suffix not given, defaults to " storage".
+	static TypePointer fromElementaryTypeName(std::string const& _name);
+
+	/// @returns boolean type.
+	static BoolType const* boolType() noexcept { return &m_boolType; }
+
+	static FixedBytesType const* byteType() { return fixedBytesType(1); }
+	static FixedBytesType const* fixedBytesType(unsigned m) { return &m_bytesM.at(m - 1); }
+
+	static ArrayType const* bytesType() noexcept { return &m_bytesType; }
+	static ArrayType const* bytesMemoryType() noexcept { return &m_bytesMemoryType; }
+	static ArrayType const* stringType() noexcept { return &m_stringType; }
+	static ArrayType const* stringMemoryType() noexcept { return &m_stringMemoryType; }
+
+	/// Constructor for a byte array ("bytes") and string.
+	static ArrayType const* arrayType(DataLocation _location, bool _isString = false);
+
+	/// Constructor for a dynamically sized array type ("type[]")
+	static ArrayType const* arrayType(DataLocation _location, Type const* _baseType);
+
+	/// Constructor for a fixed-size array type ("type[20]")
+	static ArrayType const* arrayType(DataLocation _location, Type const* _baseType, u256 const& _length);
+
+	static AddressType const* payableAddressType() noexcept { return &m_payableAddressType; }
+	static AddressType const* addressType() noexcept { return &m_addressType; }
+
+	static IntegerType const* integerType(unsigned _bits = 256, IntegerType::Modifier _modifier = IntegerType::Modifier::Unsigned)
+	{
+		solAssert((_bits % 8) == 0, "");
+		if (_modifier == IntegerType::Modifier::Unsigned)
+			return &m_uintM.at(_bits / 8 - 1);
+		else
+			return &m_intM.at(_bits / 8 - 1);
+	}
+
+	static FixedPointType const* fixedPointType(unsigned m, unsigned n, FixedPointType::Modifier _modifier);
+
+	static StringLiteralType const* stringLiteralType(std::string const& literal);
+
+	/**
+	 * @param members the member types the tuple type must contain. This is passed by value on purspose.
+	 * @returns a tuple type with the given members.
+	 */
+	static TupleType const* tupleType(std::vector<Type const*> members);
+
+	static TupleType const* emptyTupleType() noexcept { return &m_emptyTupleType; }
+
+	/// @returns a suitably simple type when a type is expected but an error has occurred.
+	static TupleType const* errorType() noexcept { return emptyTupleType(); }
```

Comment:
I'm not sure we should extend the semantics of the TypeChecker to the type system - the TypeChecker should just use `emptyTupleType`.